Realm/RLMSyncSubscription.mm

```diff
@@ -18,38 +18,144 @@
 
 #import "RLMSyncSubscription.h"
 
+#import "RLMObjectSchema_Private.hpp"
+#import "RLMObject_Private.hpp"
+#import "RLMProperty_Private.hpp"
 #import "RLMRealm_Private.hpp"
 #import "RLMResults_Private.hpp"
 #import "RLMUtil.hpp"
 
+#import "object_store.hpp"
 #import "sync/partial_sync.hpp"
 
 using namespace realm;
 
 @interface RLMSyncSubscription ()
-- (instancetype)initWithName:(NSString *)name results:(Results const&)results realm:(RLMRealm *)realm;
-
 @property (nonatomic, readwrite) RLMSyncSubscriptionState state;
 @property (nonatomic, readwrite, nullable) NSError *error;
 @end
 
 @implementation RLMSyncSubscription {
+    @public
+    NSString *_name;
+}
+
+- (instancetype)initPrivate {
+    return (self = [super init]);
+}
+
+- (void)unsubscribe {
+    __builtin_unreachable();
+}
+@end
+
+@interface RLMSyncSubscriptionObject : RLMObjectBase
+@end
+@implementation RLMSyncSubscriptionObject {
+    util::Optional<NotificationToken> _token;
+    Object _obj;
+}
+
+- (NSString *)name {
+    return _row.is_attached() ? RLMStringDataToNSString(_row.get_string(_row.get_column_index("name"))) : nil;
+}
+
+- (RLMSyncSubscriptionState)state {
+    if (!_row.is_attached()) {
+        return RLMSyncSubscriptionStateInvalidated;
+    }
+    return (RLMSyncSubscriptionState)_row.get_int(_row.get_column_index("status"));
+}
+
+- (NSError *)error {
+    if (!_row.is_attached()) {
+        return nil;
+    }
+    StringData err = _row.get_string(_row.get_column_index("error_message"));
+    if (!err.size()) {
+        return nil;
+    }
+    return [NSError errorWithDomain:RLMErrorDomain
+                               code:RLMErrorFail
+                           userInfo:@{NSLocalizedDescriptionKey: RLMStringDataToNSString(err)}];
+}
+
+- (NSString *)descriptionWithMaxDepth:(NSUInteger)depth {
+    if (depth == 0) {
+        return @"<Maximum depth exceeded>";
+    }
+
+    auto objectType = _row.get_string(_row.get_column_index("matches_property"));
+    objectType = objectType.substr(0, objectType.size() - strlen("_matches"));
+    return [NSString stringWithFormat:@"RLMSyncSubscription {\n\tname = %@\n\tobjectType = %@\n\tquery = %@\n\tstatus = %@\n\terror = %@\n}",
+            self.name, RLMStringDataToNSString(objectType),
+            RLMStringDataToNSString(_row.get_string(_row.get_column_index("query"))),
+            @(self.state), self.error];
+}
+
+- (void)unsubscribe {
+    if (_row) {
+        partial_sync::unsubscribe(Object(_realm->_realm, *_info->objectSchema, _row));
```

Comment:
In Java we implemented this as a normal managed operation, i.e. it requires a write transaction and it basically just an alias for deleting the object. I'm not sure I can what semantics make the most sense for the Cocoa binding though.

Reply:
This is something that's already part of the released API and not a new thing. I don't think the semantics we chose are great, but it's not bad enough to justify a breaking change to them.